server/apps/immich/src/api-v1/job/job.service.ts

```diff
@@ -0,0 +1,143 @@
+  metadataExtractionQueueName,
+  QueueNameEnum,
+  thumbnailGeneratorQueueName,
+  videoConversionQueueName,
+  videoMetadataExtractionProcessorName,
+} from '@app/job';
+import { InjectQueue } from '@nestjs/bull';
+import { Queue } from 'bull';
+import { BadRequestException, Inject, Injectable } from '@nestjs/common';
+import { AllJobStatusResponseDto } from './response-dto/all-job-status-response.dto';
+import { randomUUID } from 'crypto';
+import { ASSET_REPOSITORY, IAssetRepository } from '../asset/asset-repository';
+import { AssetType } from '@app/database/entities/asset.entity';
+import { GetJobDto } from './dto/get-job.dto';
+import { JobStatusResponseDto } from './response-dto/job-status-response.dto';
+
+@Injectable()
+export class JobService {
+  constructor(
+    @InjectQueue(thumbnailGeneratorQueueName)
+    private thumbnailGeneratorQueue: Queue<IThumbnailGenerationJob>,
+
+    @InjectQueue(metadataExtractionQueueName)
+    private metadataExtractionQueue: Queue<IMetadataExtractionJob>,
+
+    @InjectQueue(videoConversionQueueName)
+    private videoConversionQueue: Queue<IVideoTranscodeJob>,
+
+    @Inject(ASSET_REPOSITORY)
+    private _assetRepository: IAssetRepository,
+  ) {
+    this.thumbnailGeneratorQueue.empty();
+    this.metadataExtractionQueue.empty();
+    this.videoConversionQueue.empty();
+  }
+
+  async startJob(jobDto: GetJobDto): Promise<number> {
+    if (jobDto.jobId === QueueNameEnum.THUMBNAIL_GENERATION) {
+      const jobCount = await this.thumbnailGeneratorQueue.getJobCounts();
+
+      if (jobCount.waiting > 0) {
+        throw new BadRequestException('Thumbnail generation job is already running');
+      }
+
+      const assetsWithNoThumbnail = await this._assetRepository.getAssetWithNoThumbnail();
+
+      for (const asset of assetsWithNoThumbnail) {
+        await this.thumbnailGeneratorQueue.add(generateJPEGThumbnailProcessorName, { asset }, { jobId: randomUUID() });
+      }
+
+      return assetsWithNoThumbnail.length;
+    }
+
+    if (jobDto.jobId === QueueNameEnum.METADATA_EXTRACTION) {
+      const jobCount = await this.metadataExtractionQueue.getJobCounts();
+
+      if (jobCount.waiting > 0) {
+        throw new BadRequestException('Metadata extraction job is already running');
+      }
+
+      const assetsWithNoExif = await this._assetRepository.getAssetWithNoEXIF();
+      for (const asset of assetsWithNoExif) {
+        if (asset.type === AssetType.VIDEO) {
+          await this.metadataExtractionQueue.add(
+            videoMetadataExtractionProcessorName,
+            { asset, fileName: asset.id },
+            { jobId: randomUUID() },
+          );
+        } else {
+          await this.metadataExtractionQueue.add(
+            exifExtractionProcessorName,
+            { asset, fileName: asset.id },
+            { jobId: randomUUID() },
+          );
+        }
+      }
+      return assetsWithNoExif.length;
+    }
+
+    if (jobDto.jobId === QueueNameEnum.VIDEO_CONVERSION) {
```

Comment:
I assume this case is empty because we don't do video conversion yet? A comment indicating that (maybe with an issue reference) would be good imo.

Reply:
Ah I should leave some comment in there. It will be implemented once the PR for video setting is merged